File: python/sml/methods/apportionment.py
from pyspark.sql import DataFrame


class Apportionment:

    def __init__(self, df=None):

        if df is None:
            raise TypeError

        self._df = df

        self._jApp = self._df._sc._jvm                                          \
                         .uk.gov.ons                                            \
                         .api.java.methods                                      \
                         .JavaApportionmentFactory.apportionment(self._df._jdf)

    def __mandatory_argument_check(self, *args):
        for arg in args:
            if arg is None:
                raise TypeError

    def app1(self, aux_col=None, date_col=None,
             agg_col_a=None, agg_col_b=None, app_cols=None):

        """
        A Python wrapper for the Apportionment function.

        :param aux_col: Column used as a weighting.
        :param date_col: Column used to differentiate records of the same reference.
        :param agg_col_a: Column containing initial reference.
        :param agg_col_b: Column containing reference to apportion to.
        :param app_cols: Column(s) to be redistributed.
        :return: pyspark.sql.DataFrame
        """

        self.__mandatory_argument_check(aux_col, date_col, agg_col_a, agg_col_b, app_cols)

        df = self._df

        return DataFrame(self._jApp.app1(aux_col, date_col,
                                         agg_col_a, agg_col_b, app_cols),
                         df.sql_ctx)


def apportionment(df):

    return Apportionment(df)

File: src/main/scala/uk/gov/ons/methods/Apportionment.scala
package uk.gov.ons.methods

import uk.gov.ons.methods.impl.ApportionmentImpl._
import org.apache.spark.sql.{DataFrame, Dataset, Row}

class Apportionment (val df: DataFrame) {

  if (df == null) throw new Exception("DataFrame cannot be null")

  val defaultCol = "AppResult"

  private def mandatoryArgCheck(arg1: String, arg2: String, arg3: String, arg4: String): Unit = {
    if ((arg1 == null) || (arg2 == null) || (arg3 == null) || (arg4 == null)) throw new Exception("Missing mandatory argument")
  }

  /**
    * Scala method that calls the Apportionment implicit method.
    *
    * @author danny.fiske@ons.gov.uk
    * @version 1.0
    *
    * @param auxColumn    String - column used as a weighting.
    * @param dateColumn   String - column used to differentiate records of the same reference.
    * @param aggColumnA   String - column containing initial reference.
    * @param aggColumnB   String - column containing reference to apportion to.
    * @param appColumns   List[String] - column(s) to be redistributed.
    * @return             Dataset[Row]
    */
  def app1(auxColumn: String, dateColumn: String, aggColumnA:
  String, aggColumnB: String, appColumns: List[String]): DataFrame = {

    mandatoryArgCheck(auxColumn, dateColumn, aggColumnA, aggColumnB)

    val dF = if (df == null) df else df

    dF.checkColNames(auxColumn, dateColumn, aggColumnA, aggColumnB)
      .app1(auxColumn, dateColumn, aggColumnA, aggColumnB, appColumns)
  }
}

object Apportionment {

  def apportionment(df: DataFrame) : Apportionment = new Apportionment(df)
}
